internal/lsp/server.go

```diff
@@ -0,0 +1,522 @@
+				WorkspaceDiagnostics:  true,
+			},
+			Workspace: WorkspaceOptions{
+				FileOperations: FileOperationsServerCapabilities{
+					DidCreate: FileOperationRegistrationOptions{
+						Filters: []FileOperationFilter{regoFilter},
+					},
+					DidRename: FileOperationRegistrationOptions{
+						Filters: []FileOperationFilter{regoFilter},
+					},
+					DidDelete: FileOperationRegistrationOptions{
+						Filters: []FileOperationFilter{regoFilter},
+					},
+				},
+			},
+		},
+	}
+
+	folderURI, err := url.Parse(l.clientRootURI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	// load the rego source files into the cache
+	err = filepath.Walk(folderURI.Path, func(path string, info os.FileInfo, err error) error {
+		stat, err := os.Stat(path)
+		if err != nil {
+			return fmt.Errorf("failed to stat %q: %w", path, err)
+		}
+		if stat.IsDir() {
+			return nil
+		}
+
+		// TODO(charlieegan3): make this configurable for things like .rq etc?
+		if !strings.HasSuffix(path, ".rego") {
+			return nil
+		}
+
+		_, _, err = updateCacheForURIFromDisk(l.cache, fmt.Sprintf("file://%s", path))
+		if err != nil {
+			return fmt.Errorf("failed to update cache for uri %q: %w", path, err)
+		}
+
+		_, err = updateParse(l.cache, fmt.Sprintf("file://%s", path))
+		if err != nil {
+			return fmt.Errorf("failed to update parse: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to walk workspace dir %q: %w", folderURI.Path, err)
+	}
+
+	// attempt to load the config as it is found on disk
+	file, err := config.FindConfig(strings.TrimPrefix(l.clientRootURI, "file://"))
+	if err == nil {
+		l.reloadConfig(file, false)
+	}
+
+	l.logOutboundMessage("initialize", result)
+
+	return result, nil
+}
+
+func (l *LanguageServer) reloadConfig(configReader io.Reader, runWorkspaceDiagnostics bool) {
+	l.loadedConfigLock.Lock()
+	defer l.loadedConfigLock.Unlock()
+
+	var loadedConfig config.Config
+	err := yaml.NewDecoder(configReader).Decode(&loadedConfig)
+	if err != io.EOF && err != nil {
+		l.logError(fmt.Errorf("failed to reload config: %w", err))
+		return
+	}
+
+	// if the config is now blank, then we need to clear it
+	if err == io.EOF {
+		l.loadedConfig = nil
+	} else {
+		l.loadedConfig = &loadedConfig
+	}
+
+	// this can be set to false by callers to disable the running of diagnostics for the whole workspace.
+	// this is intended to be used at start up when a workspace run is already going to be taking place.
+	if runWorkspaceDiagnostics {
+		l.diagnosticRequestWorkspace <- fmt.Sprintf("config file changed")
+	}
+}
+
+func (l *LanguageServer) handleInitialized(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) (result interface{}, err error) {
+
+	l.diagnosticRequestWorkspace <- "initialized"
+
+	return nil, nil
+}
+
+func (l *LanguageServer) handleTextDocumentDiagnostic(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) (result interface{}, err error) {
```

Comment:
[nit]
```suggestion
func (*LanguageServer) handleTextDocumentDiagnostic(context.Context, *jsonrpc2.Conn, *jsonrpc2.Request) (any, error) {
```

Reply:
Hmm, I wonder if the linter could be configured for this preference?

Reply:
https://github.com/StyraInc/regal/pull/534 Yup!

Reply:
Thanks, this should be all fixed now!